video/vdu_buffered.h

```diff
@@ -1060,4 +1072,203 @@ void VDUStreamProcessor::bufferCopyAndConsolidate(uint16_t bufferId, std::vector
+		hdr.marker[2] = 'p';
+		hdr.type = COMPRESSION_TYPE_TURBO;
+		hdr.orig_size = 0;
+
+		auto p_hdr_bytes = hdr.marker;
+		for (int i = 0; i < sizeof(hdr); i++) {
+			local_write_compressed_byte(&cd, *p_hdr_bytes++);
+		}
+
+		// loop thru blocks stored against the source buffer ID
+		uint32_t orig_size = 0;
+		for (const auto &block : buffers[sourceBufferId]) {
+			// compress the block into our temporary buffer
+			auto bufferLength = block->size();
+			auto p_data = block->getBuffer();
+			orig_size += bufferLength;
+			debug_log(" from buffer %u [%08X] %u bytes\n\r", sourceBufferId, p_data, bufferLength);
+			debug_log(" %02hX %02hX %02hX %02hX\n\r",
+						p_data[0], p_data[1], p_data[2], p_data[3]);
+			cd.input_count += bufferLength;
+			while (bufferLength--) {
+				agon_compress_byte(&cd, *p_data++);
+			}
+		}
+		agon_finish_compression(&cd);
+
+		// make a single buffer with all of the temporary output data
+		auto bufferStream = make_shared_psram<BufferStream>(cd.output_count);
+		if (!bufferStream || !bufferStream->getBuffer()) {
+			// buffer couldn't be created
+			debug_log("bufferCompress: failed to create buffer %d\n\r", bufferId);
+			return;
+		}
+
+		auto p_hdr = (CompressionFileHeader*) p_temp;
+		p_hdr->orig_size = orig_size;
+
+		auto destination = bufferStream->getBuffer();
+		debug_log(" %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX %02hX\n\r",
+					p_temp[0], p_temp[1], p_temp[2], p_temp[3],
+					p_temp[4], p_temp[5], p_temp[6], p_temp[7],
+					p_temp[8], p_temp[9], p_temp[10], p_temp[11]);
+		memcpy(destination, p_temp, cd.output_count);
+		heap_caps_free(p_temp);
+		buffers[bufferId].push_back(bufferStream);
+
+		uint32_t pct = (cd.output_count * 100) / cd.input_count;
+		debug_log("Compressed %u input bytes to %u output bytes (%u%%) at %08X\n\r",
+				cd.input_count, cd.output_count, pct, destination);
+	} else {
+		debug_log("bufferCompress: cannot allocate temporary buffer of %d bytes\n\r", COMPRESSION_OUTPUT_CHUNK_SIZE);
+	}
+}
+
+// VDU 23, 0, &A0, bufferId; &1D, sourceBufferId; sourceBufferId; ...; 65535; : Decompress blocks from buffers
+// Decompress (blocks from) a list of buffers into a new buffer.
+// List is terminated with a bufferId of 65535 (-1).
+// Replaces the target buffer with the new one.
+// If target buffer is included in the source list it will be skipped.
+//
+void VDUStreamProcessor::bufferDecompress(uint16_t bufferId, uint16_t sourceBufferId) {
+	// Validate the compression header
+	if (buffers[sourceBufferId][0]->size() < sizeof(CompressionFileHeader)) {
+		debug_log("bufferDecompress: buffer too small for header\n\r");
+		return;
+	}
+	
+	auto p_hdr = (const CompressionFileHeader*) buffers[sourceBufferId][0]->getBuffer();
+	if (p_hdr->marker[0] != 'C' ||
+		p_hdr->marker[1] != 'm' ||
+		p_hdr->marker[2] != 'p' ||
+		p_hdr->type != COMPRESSION_TYPE_TURBO) {
+		debug_log("bufferDecompress: header is invalid\n\r");
+		return;
+	}
+	auto orig_size = p_hdr->orig_size;
```

Comment:
this could maybe be refactored into a `getLengthFromCompressionHeader` function (or something like that), which could return `0` if an invalid header is found

it could also be potentially called in the loop below, that way every block in the buffer would get its header checked rather than just skipped

Reply:
The assumption there being that we would use this to decompress multiple blocks, each with a separate header? I hadn't really considered that as a possibility.

Reply:
each block already has a header tho - the loop below that's iterating over blocks in a buffer is skipping the header for each block

and yeah this could potentially allow for funky things in the future, mixing and matching compression types - fun times 😉

Reply:
actually in hindsight (whilst further looking at this code and trying to refactor a little bit) I misunderstood a little about what the code was doing

my assertion "each block already has a header" wasn't correct.  I'd got confused by the use of `skip_hdr`, and hadn't noticed that value was getting zero'd after the first block in a buffer had been read/processed

the code only assumes that the first block is has a header, and that subsequent blocks in a buffer form part of the same chunk of compressed data.  from the perspective of sending data over from MOS to the VDP in separate chunks this makes perfect sense.  _without_ this functionality, chunks would need to be consolidated before they could be decompressed

I think there's potentially still some merit in the idea of having multiple blocks with separate headers.  one use of this could be for the compressor to pick different compression algorithms for different sections of data.  another use could be to compose chunks of compressed data via the "copy references" command for later decompression.

I'll have a think about whether/how to implement this idea